### PR TITLE
DROID-2307 Notifications | Enhancement | Intercept notifications related to gallery experience installation

### PR DESCRIPTION
**Pull request by konstantiniiv**
---

- [x] I understand that contributing to this repository will require me to agree with the [CLA](https://github.com/anyproto/.github/blob/main/docs/CLA.md)

---

### What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI

### Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

### Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 [tech-docs](https://github.com/anyproto/tech-docs)
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: presentation/src/main/java/com/anytypeio/anytype/presentation/home/HomeScreenViewModel.kt [hunk 1]
```diff
@@ -163,7 +165,8 @@ class HomeScreenViewModel(
     private val getSpaceView: GetSpaceView,
     private val searchObjects: SearchObjects,
     private val getPinnedObjectTypes: GetPinnedObjectTypes,
-    private val userPermissionProvider: UserPermissionProvider
+    private val userPermissionProvider: UserPermissionProvider,
+    private val notificationsChannel: NotificationsChannel
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/home/HomeScreenViewModel.kt:169]: no longer needed?

File: presentation/src/main/java/com/anytypeio/anytype/presentation/home/HomeScreenViewModel.kt [hunk 2]
```diff
@@ -202,6 +205,8 @@ class HomeScreenViewModel(
 
     val hasEditAccess = userPermissions.map { it?.isOwnerOrEditor() == true }
 
+    val notifications : MutableStateFlow<Notification?> = MutableStateFlow(null)
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/home/HomeScreenViewModel.kt:208]: no longer needed?

File: presentation/src/main/java/com/anytypeio/anytype/presentation/notifications/NotificationsViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+class NotificationsViewModel(
+    private val analytics: Analytics,
+    private val notificationsProvider: NotificationsProvider,
+    private val spaceManager: SpaceManager,
+    private val saveCurrentSpace: SaveCurrentSpace
+) : BaseViewModel() {
+
+    val state = MutableStateFlow<NotificationsScreenState>(NotificationsScreenState.Hidden)
+    val command = MutableStateFlow<Command?>(null)
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/notifications/NotificationsViewModel.kt:26]: MutableSharedFlow?

File: presentation/src/main/java/com/anytypeio/anytype/presentation/notifications/NotificationsViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                }
+            }
+        }
+    }
+
+    fun onErrorButtonClick() {
+        command.value = Command.Dismiss
+    }
+
+    fun onSpaceClick(spaceId: SpaceId) {
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/notifications/NotificationsViewModel.kt:57]: onNavigateToSpaceClicked

File: presentation/src/main/java/com/anytypeio/anytype/presentation/notifications/NotificationsProvider.kt [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+interface NotificationsProvider {
+    fun start()
+    fun stop()
+    fun get(): Notification?
+    fun observe(): Flow<Notification?>
+
+    class Default @Inject constructor(
+        private val dispatchers: AppCoroutineDispatchers,
+        private val scope: CoroutineScope,
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/notifications/NotificationsProvider.kt:24]: are you using app scope here?

**Review by uburoiubu — APPROVED** [review]

File: presentation/src/main/java/com/anytypeio/anytype/presentation/home/HomeScreenViewModel.kt (hunk 1 above)
Comment (konstantiniiv) [reply presentation/src/main/java/com/anytypeio/anytype/presentation/home/HomeScreenViewModel.kt:169]: yes, all Notifications logic moved to Main View Model for now

File: presentation/src/main/java/com/anytypeio/anytype/presentation/home/HomeScreenViewModel.kt (hunk 2 above)
Comment (konstantiniiv) [reply presentation/src/main/java/com/anytypeio/anytype/presentation/home/HomeScreenViewModel.kt:208]: yes

File: presentation/src/main/java/com/anytypeio/anytype/presentation/notifications/NotificationsProvider.kt (hunk 5 above)
Comment (konstantiniiv) [reply presentation/src/main/java/com/anytypeio/anytype/presentation/notifications/NotificationsProvider.kt:24]: @Named(ConfigModule.DEFAULT_APP_COROUTINE_SCOPE) scope: CoroutineScope, this one

File: presentation/src/main/java/com/anytypeio/anytype/presentation/notifications/NotificationsViewModel.kt (hunk 4 above)
Comment (konstantiniiv) [reply presentation/src/main/java/com/anytypeio/anytype/presentation/notifications/NotificationsViewModel.kt:57]: done

File: presentation/src/main/java/com/anytypeio/anytype/presentation/home/HomeScreenViewModel.kt (hunk 2 above)
Comment (konstantiniiv) [reply presentation/src/main/java/com/anytypeio/anytype/presentation/home/HomeScreenViewModel.kt]: deleted

File: presentation/src/main/java/com/anytypeio/anytype/presentation/notifications/NotificationsProvider.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        private val notificationsChannel: NotificationsChannel,
+        private val awaitAccountStartManager: AwaitAccountStartManager
+    ) : NotificationsProvider {
+
+        @OptIn(ExperimentalCoroutinesApi::class)
+        override fun observe(): Flow<List<Notification.Event>> {
+            return combine(
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/notifications/NotificationsProvider.kt:28]: or:

```
awaitAccountStartManager.isStarted().flatMapLatest { isStarted -> 
if (isStarted) notificationsChannel.observe() else emptyFlow()
}
```

+ please add test verifying this behavior:
started, one event sent, one event emitted, stopped, one event sent, no event emitted, etc.

Comment (konstantiniiv) [reply presentation/src/main/java/com/anytypeio/anytype/presentation/notifications/NotificationsProvider.kt]: updated + added tests

**Pull request merged**